### ISSUE TITLE
validate returns true although nested property has AMV error

### ISSUE THREAD

**Issue by sliminas** [trailblazer/reform-rails#50]
I created a test to show the unexpected behavior in https://github.com/sliminas/reform-rails/commit/9bf826000cfe3e800fd6bd3c8ae90c5d63562019

I already tracked down the issue to this line in the code https://github.com/trailblazer/reform-rails/blob/master/lib/reform/form/active_model/validations.rb#L62

```ruby
def validate!(params, pointers=[])
  @amv_errors = ActiveModel::Errors.new(self)

  super.tap do
    # @fran: super ugly hack thanks to the shit architecture of AMV. let's drop it in 3.0 and move on!
    all_errors = @result.instance_variable_get(:@results)

    @result = Reform::Contract::Result.new(all_errors)
```

If you look at @result at the first time we arrive there during the test it looks like this and all_errors will be the AMV error message:
```ruby
#<Reform::Contract::Result:0x0055faa3f42a10 @failure={:length=>["must be greater than 55"]}, @results=[{:length=>["must be greater than 55"]}]>
```

But we get there a second time later one, where `@result` looks like this:
```ruby
#<Reform::Contract::Result:0x0055b0a4c3ad98
 @failure=
  #<Reform::Contract::Result:0x0055b0a4c3b400 @failure={:length=>["must be greater than 55"]}, @results=[{:length=>["must be greater than 55"]}]>,
 @results=[]>
```
But there the instance variable @results is empty and the success? method will return true, although there are errors present in the contract.

I used ruby 2.3.1 for the test.

**Comment by apotonick** [comment]
Thanks for tracking this down! Why do we get there a second time, I wonder?

**Comment by sliminas** [comment]
After some more debugging I know this so far:

We get there twice because in the test we have a nested form, so the first time of validate! is for the nested form for the 'hit' property and the second time of validate! is executed for the PopulatedAlbumForm itself.
I think this works as intended.

The actual problem seems to be that a new `Reform::Contract::Result` is built with only reading the `@results` instance variable, although there are just the local results in this variable and not the possibly present nested_results of a nested form with errors (https://github.com/trailblazer/reform/blob/v2.3.0.rc1/lib/reform/result.rb#L9) :
```ruby
class Result
  def initialize(results, nested_results=[]) # DISCUSS: do we like this?
    @results = results # native Result objects, e.g. `#<Dry::Validation::Result output={:title=>"Fallout", :composer=>nil} errors={}>`
    @failure = (results + nested_results).find(&:failure?) # TODO: test nested.
```

The second time we are at line 62 `@result` looks like this:
```ruby
#<Reform::Contract::Result:0x0055ed326e05f8
 @failure=
  #<Reform::Contract::Result:0x0055ed326e1278 @failure={:length=>["must be greater than 55"]}, @results=[{:length=>["must be greater than 55"]}]>,
 @results=[]>
```

Now a new result is built with `@result = Reform::Contract::Result.new(all_errors)` where `all_errors` is an empty array since `@result`'s `@results` variable is empty.
So I think there should be sth. like this happening, since the nested errors are only present in the `@failure` variable:
```ruby
all_errors = @result.instance_variable_get(:@results)
nested_errors = @result.instance_variable_get(:@failure)
@result = Reform::Contract::Result.new(all_errors, [nested_errors].compact)
```
This seems to work for me, see:
https://github.com/sliminas/reform-rails/commit/d1b535a2fc4b63b8cb359f77a85a846b963ef266

**Comment by fernandes** [comment]
Closing as was fixed in #53
